File: Flink_WuShengran/src/main/scala/com/atguigu/wc/StreamWordCount.scala
package com.atguigu.wc

import org.apache.flink.api.java.utils.ParameterTool
import org.apache.flink.streaming.api.scala._

/**
 * Created by John.Ma on 2020/9/10 0010 22:36
 * 流处理Word Count
 */
object StreamWordCount {
    def main(args: Array[String]): Unit = {
        //创建流处理的执行环境
        val env: StreamExecutionEnvironment = StreamExecutionEnvironment.getExecutionEnvironment

        env.setParallelism(1)

        //接收一个socket文本流
        val inputDataStream = env.socketTextStream("172.16.197.86", 4444)

        inputDataStream
            .flatMap(_.split(" "))
            .filter(_ nonEmpty)
            .map((_,1))
            .keyBy(_._1)
            .sum(1)
            .print()

        //启动任务执行
        env.execute("stream word count")

        //结果解释: 2> 前面的数字是一个当前统计的,当前并行任务的号码,4个线程,多线程
        //多线程并行度是4,当前机器的核数
        //2> (hello,16)
        //2> (hello,17)
        //2> (hello,18)
        //2> (hello,19)
        //2> (hello,20)

    }

}

File: Flink_WuShengranJava/src/main/java/com/atguigu/wc/StreamWordCount.java
package com.atguigu.wc;

import org.apache.flink.api.common.functions.RichFlatMapFunction;
import org.apache.flink.api.java.tuple.Tuple2;
import org.apache.flink.streaming.api.datastream.DataStream;
import org.apache.flink.streaming.api.environment.StreamExecutionEnvironment;
import org.apache.flink.util.Collector;

/**
 * Created by John.Ma on 2021/6/26 0026 23:43
 */
public class StreamWordCount {
    public static void main(String[] args) throws Exception{
        //创建流处理执行环境
        StreamExecutionEnvironment env = StreamExecutionEnvironment.getExecutionEnvironment();
        // 从文件里面读取
        // 从文件中读取器数据
//        DataStream<String> inputStream = env.readTextFile("D:\\fileImportant\\Learn_projects\\learn\\Flink_WuShengranJava\\src\\main\\resources\\hello.txt");

        DataStream<String> inputStream = env.socketTextStream("www.bigdata01.com",7777);
        inputStream.flatMap(new RichFlatMapFunction<String, Tuple2<String,Integer>>() {
            @Override
            public void flatMap(String value, Collector<Tuple2<String, Integer>> out) throws Exception {
                String[] arr = value.split(" ");
                for(String s1:arr){
                    out.collect(Tuple2.of(s1,1));
                }
            }
        }).keyBy(tuple2 -> tuple2.f0)
                .sum(1)
                .print();




        env.execute();


    }
}
